src/singleprover.cpp

```diff
@@ -123,14 +123,21 @@ json SingleProver::startProve(std::string input)
     std::string output("Witness generation finished in " + std::to_string(duration) + "ms");
     LOG_INFO(output);
 
-    auto proof = prover->prove(wtnsData);
-    auto t2 = std::chrono::steady_clock::now();
-    duration = std::chrono::duration_cast<std::chrono::milliseconds>(t2 - t1).count();
-    output = "Proof generation finished in " + std::to_string(duration) + "ms";
-    LOG_INFO(output);
-
     unlink(witnessFileName.c_str());
     unlink(inputFileName.c_str());
 
+    return prove(wtnsData);
+}
+
+json SingleProver::prove(AltBn128::FrElement *wtnsData) {
+    // The mutex is set because the performance of prover->prove degrades significantly
+    //  when multiple instances are run in parallel.
+    std::lock_guard<std::mutex> guard(mtx);
```

Comment:
put the lock on the smallest block possible (just prover->prove())

Reply:
I have moved toJSON out. Rest of the code is for measuring time taken.